packages/cdk8s-plus-17/src/statefulset.ts

```diff
@@ -0,0 +1,217 @@
+import { ApiObject, ApiObjectMetadataDefinition, Lazy, Names } from 'cdk8s';
+import { Construct } from 'constructs';
+import { Resource, ResourceProps } from './base';
+import { Container, ContainerProps } from './container';
+import * as k8s from './imports/k8s';
+import { RestartPolicy, PodTemplate, IPodTemplate, PodTemplateProps } from './pod';
+import { Service, ServiceProps, ServiceType } from './service';
+import { IServiceAccount } from './service-account';
+import { Volume } from './volume';
+
+
+export enum PodManagementPolicy {
+  ORDERED_READY = 'OrderedReady',
+  PARALLEL = 'Parallel',
+}
+
+/**
+ * Properties for initialization of `StatefulSet`.
+ */
+export interface StatefulSetProps extends ResourceProps, PodTemplateProps {
+  /**
+    * Number of desired pods.
+    *
+    * @default 1
+    */
+  readonly replicas?: number;
+
+  /**
+    * Automatically allocates a pod selector for this statefulset.
+    *
+    * If this is set to `false` you must define your selector through
+    * `statefulset.podMetadata.addLabel()` and `statefulset.selectByLabel()`.
+    *
+    * @default true
+    */
+  readonly defaultSelector?: boolean;
+
+  /**
+    * Pod management policy to use for this statefulset.
+    *
+    * @default PodManagementPolicy.ORDERED_READY
+    */
+  readonly podManagementPolicy?: PodManagementPolicy;
+
+  /**
+    * Details of service to create and expose.
+    *
+    * @default undefined - allocate a default service based upon servicePort and serviceName.
+    */
+  readonly service?: ServiceProps;
```

Comment:
We can just accept a `Service` here instead, it will prevent some logic duplication and keep things more modular. Its ok to require the caller to instantiate other constructs.

Reply:
I thought about that but it ends up making the initialization more complex.  I was trying to get an API as easy to use as the deployment expose() call.

For example when I am using this in my production chart I end up calling as:

```
const redis_set = new kplus.StatefulSet(this, 'redis-set', {
         serviceName: 'redis-srv',
         replicas: 2,
         containers: [redis_container],
         servicePort: redis_port,
         service: {
            type: kplus.ServiceType.CLUSTER_IP,
            clusterIP: undefined, // force it to allocate
         },
      });
```

If the user had to allocate the Service it would look like:

```
const redis_srv = new Service(this, 'RedisService', {
   metadata: {name: 'redis-srv'},
   type: ServiceType.CLUSTER_IP,
   clusterIP: undefined,
   ports: [{port: redis_port}],
})

const redis_set = new kplus.StatefulSet(this, 'redis-set', {
         serviceName: 'redis-srv',
         replicas: 2,
         containers: [redis_container],
         servicePort: redis_port,
         service: redis_srv
      });
```
And the callee would have to make sure the service name strings match and the ports match.  Seems like this is the type of thing we would want to wrap to make it simpler to handle and then if they need to modify the service they can patch it.

What do you think though?  I probably missed something in this one.